app/scripts/analysis/charts/charts.coffee

```diff
@@ -1,86 +1,671 @@
+            obj.push(JSON.parse(tmp))
+            i++
+
+        console.log obj
+        return obj
+      console.log $scope.graphInfo
+      send = graphFormat()
+      results = {
+        data: send
+        xLab: _headers[$scope.graphInfo.x],
+        yLab: _headers[$scope.graphInfo.y],
+        zLab: _headers[$scope.graphInfo.z],
+        name: $scope.graphInfo.graph
+      }
+      #_graphInfo = results
+      #[_chartData[$scope.graphInfo.x], _chartData[$scope.graphInfo.y]]
+      $rootScope.$broadcast 'charts:graphDiv', results
+
+    $scope.changeName = ()->
+      $scope.graphInfo.graph = $scope.graphSelect.name
+      $scope.createGraph()
+
+    $scope.changeVar = (selector,headers, ind) ->
+      for h in headers
+        if selector.value is h.value then $scope.graphInfo[ind] = parseFloat h.key
+      $scope.createGraph()
+
+    sb = ctrlMngr.getSb()
+
+    # deferred = $q.defer()
+
+    token = sb.subscribe
+      msg:'take table'
+      msgScope:['charts']
+      listener: (msg, _data) ->
+        _headers = d3.entries _data.header
+        $scope.headers = _headers
+        _chartData = dataTransform.format(_data.data)
+
+    sb.publish
+      msg:'get table'
+      msgScope:['charts']
+      callback: -> sb.unsubscribe token
+      data:
+        tableName: $stateParams.projectId + ':' + $stateParams.forkId
+
+
+])
+
+.factory('app_analysis_charts_dataTransform',[
+  () ->
+
+    _transpose = (data) ->
+      data[0].map (col, i) -> data.map (row) -> row[i]
+
+    _transform = (data) ->
+      for col in data
+        obj = {}
+        for value, i in col
+          obj[i] = value
+        d3.entries obj
+
+    _format = (data) ->
+      return _transform(_transpose(data))
+
+    transform: _transform
+    transpose:_transpose
+    format: _format
+
+])
+
+
+.directive 'd3Charts', [
+  () ->
+    restrict:'E'
+    template:"<div class='graph-container' style='height: 600px'></div>"
+    link: (scope, elem, attr) ->
+      margin = {top: 10, right: 40, bottom: 30, left:40}
+      width = 900 - margin.left - margin.right
+      height = 500 - margin.top - margin.bottom
+      x = null
+      y = null
+      xMax = null
+      yMax = null
+      xMin = null
+      yMin = null
+      xAxis = null
+      yAxis = null
+      svg = null
+      data = null
+      _graph = null
+      container = null
+      pieData = null
+      gdata = null
+
+      makepieData = (data) ->
+        pieMax = d3.max(data, (d)->parseFloat d.x)
+        pieMin = d3.min(data, (d)->parseFloat d.x)
+        a = 0
+        b = 0
+        c = 0
+        d = 0
+        e = 0
+        f = 0
+        g = 0
+        rangeInt = (pieMax - pieMin)/7
+#        console.log pieMin+rangeInt
+        for da in data
+          val = parseFloat da.x
+
+#          console.log val, pieMin+rangeInt
+
+          if val < (pieMin+rangeInt)
```

Comment:
(refactor): if I correctly understand what are you doing here, this is not very elegant implementation of search and sorting of values, you can do something like this:
```
maxPiePieces = 7  # set magic constant to variable
xPoints = data.map(el) -> el.x
piePieces = [0..maxPiePieces - 1]  # create array with numbers of pie pieces
pieceVolumes = pieces.map(pieceNum) ->
  currPiecePoints = xPoints.filter(x) -> pieMin + pieceNum * rangeInt <= x < pieMin + (pieceNum + 1) * rangeInt  # filter data points by limits of each piece
  currPiecePoints.length  # return number of elements in each piece
```